src/Spice86.Core/Emulator/Memory/Memory.cs

```diff
@@ -1,27 +1,259 @@
+    /// </summary>
+    public const uint ConvMemorySize = 1024 * 1024;
+
+    public bool EnableA20 { get; internal set; }
+
+    public int MemorySize { get; init; }
+
     private readonly BreakPointHolder _readBreakPoints = new();
 
     private readonly BreakPointHolder _writeBreakPoints = new();
 
+    private readonly MetaAllocator _metaAllocator = new();
+
     // For breakpoints to access what is getting written
     public byte CurrentlyWritingByte { get; private set; } = 0;
 
-    public Memory(uint sizeInKb) {
+    private readonly Machine _machine;
+
+    public Memory(Machine machine, uint sizeInKb) {
+        if (sizeInKb * 1024 < ConvMemorySize) {
+            throw new ArgumentException("Memory size must be at least 1 MB.");
+        }
+        this.MemorySize = (int)sizeInKb * 1024;
+
+        // Reserve room for the real-mode interrupt table.
+        this.Reserve(0x0000, 256 * 4);
+
+        // Reserve VGA video RAM window.
+        this.Reserve(0xA000, VramUpperBound - VramAddress + 16u);
+
+        _machine = machine;
         Ram = new byte[sizeInKb * 1024];
+        unsafe {
+            fixed(byte* ramPtr = this.Ram) {
+                RawView = ramPtr;
+            }
+        }
         UInt8 = new(this);
         UInt16 = new(this);
         UInt32 = new(this);
     }
 
+    /// <summary>
+    /// Reserves a block of conventional memory.
+    /// </summary>
+    /// <param name="minimumSegment">Minimum segment of requested memory block.</param>
+    /// <param name="length">Size of memory block in bytes.</param>
+    /// <returns>Information about the reserved block of memory.</returns>
+    public ReservedBlock Reserve(ushort minimumSegment, uint length) {
+        ushort allocation = _metaAllocator.Allocate(minimumSegment, (int)length);
+        return new(allocation, length);
+    }
+
+    private ushort _nextHandlerOffset = 4096;
+
+    /// <summary>
+    /// Segment for interrupt/callback proxies.
+    /// </summary>
+    private const ushort HandlerSegment = 0xF100;
+
+    /// <summary>
+    /// Writes a new callback handler and returns its address.
+    /// </summary>
+    /// <param name="id">Unique ID of the callback handler.</param>
+    /// <param name="hookable">Value indicating whether the callback is hookable.</param>
+    /// <returns>Address of the callback handler.</returns>
+    internal SegmentedAddress AddCallbackHandler(byte id, bool hookable) {
+        IntPtr ptr = GetPointer(HandlerSegment, _nextHandlerOffset);
+        int length = 4;
+        unsafe {
+            byte* writePtr = (byte*)ptr.ToPointer();
+            if (hookable) {
+                writePtr[0] = 0xE9; // JMP iw
+                writePtr[1] = 3; // jump past the 3 nops
+                writePtr[2] = 0;
+                writePtr[3] = 0x90;
+                writePtr[4] = 0x90;
+                writePtr[5] = 0x90;
+
+                writePtr += 6;
+                length += 6;
+            }
+
+            writePtr[0] = 0x0F;
```

Comment:
Our CPU will not understand this. We use opcode FE38 which is the same dosbox uses.

Reply:
Fixed in feature/xms